pkg/query-service/app/querier/v2/helper.go

```diff
@@ -138,13 +210,36 @@ func (q *querier) runBuilderQuery(
 				}
 				return
 			}
+
+			filteredSeries, startTime, endTime := filterSeriesPoints(series, miss.Start, miss.End, builderQuery.StepInterval)
+
+			// making sure that empty range doesn't doesn't enter the cache
+			if len(filteredSeries) > 0 {
+				filteredMissedSeries = append(filteredMissedSeries, querycache.CachedSeriesData{
+					Data:  filteredSeries,
+					Start: startTime,
+					End:   endTime,
+				})
+			}
+
+			// for the actual response
 			missedSeries = append(missedSeries, querycache.CachedSeriesData{
 				Data:  series,
 				Start: miss.Start,
 				End:   miss.End,
 			})
 		}
-		mergedSeries := q.queryCache.MergeWithCachedSeriesData(cacheKeys[queryName], missedSeries)
+
+		filteredMergedSeries, err := q.queryCache.MergeWithCachedSeriesDataV2(cacheKeys[queryName], filteredMissedSeries)
+		if err != nil {
+			filteredMergedSeries = filteredMissedSeries
+		}
+		q.queryCache.StoreSeriesInCache(cacheKeys[queryName], filteredMergedSeries)
+
+		mergedSeries, err := q.queryCache.MergeWithCachedSeriesDataV2(cacheKeys[queryName], missedSeries)
+		if err != nil {
+			mergedSeries = missedSeries
+		}
 
 		resultSeries := common.GetSeriesFromCachedData(mergedSeries, start, end)
```

Comment:
The logic in the function `GetSeriesFromCachedData` filters out the items that have timestamp < start but it's possible that the first point has timestamp < start which is not returned to client?

Reply:
yeah, the getSeriesFromCache data actually filters that item, thanks for the catch.